### PR TITLE
Multiple issues: Fixed errors with updating optional profile fields

### PR DESCRIPTION
**Pull request by katelynienaber**
## Proposed changes

This is one change that fixes several issues.

1. **Fix for issue #989:** There has been a bugfix! 😃 Which means in Zowe Explorer we can now pass an `IUpdateProfile` object to `CliProfileManager.update()`...and we don't have to change the profile object key names to resemble CLI command params. This was the cause of #989.

2. **Fix for issue #1095:** I set `merge: false` in the profile, so only the new values will be kept, not the old (ie. now we don't have to store an empty string as a placeholder for an optional field).

3. **Fix for issue #1016:** There are two new conditions in our `updateProfile()` function, which will remove the `responseTimeout` field if it was set to 0, and any other field if it was set to an empty string. And because of fix 2 (above), #1016 should not reoccur.

Linked issues (bugs):
[#989](https://github.com/zowe/vscode-extension-for-zowe/issues/989)
[#1016](https://github.com/zowe/vscode-extension-for-zowe/issues/1016)
[#1095](https://github.com/zowe/vscode-extension-for-zowe/issues/1095)

Other linked issues (to-dos that are no longer needed):
[#1038](https://github.com/zowe/vscode-extension-for-zowe/issues/1038)
[#1090](https://github.com/zowe/vscode-extension-for-zowe/issues/1090)

## Release Notes

Milestone: 1.12
Changelog: Fixed errors with updating optional profile fields

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [x] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [x] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [x] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

**Comment by JillieBeanSim** [comment]
Hey @katelynienaber I was testing this PR, while creating a profile I still get `""` for optional values left blank along with the `0` value for encoding and response timeout that where left blank/entered passed.  
image

Then I edited the same profile leaving those same values blank and still got `0` in the yaml for them.
image

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]
see comment above for issues I ran into.

the changes requested are:
Should we use this same way of thinking during create?
Also, would like to see the default value for encoding used or leave it out of the yaml 
Finally, shouldn't it leave response timeout out of the yaml if value = 0?

**Comment by katelynienaber** [comment]
> see comment above for issues I ran into.
> 
> the changes requested are:
> Should we use this same way of thinking during create?
> Also, would like to see the default value for encoding used or leave it out of the yaml
> Finally, shouldn't it leave response timeout out of the yaml if value = 0?

Hey @JillieBeanSim, give it a shot now.

**Should we use this same way of thinking during create?**
I think so. Create is kind of hacky right now because the `CliProfileManager.save` function does some validation, and won't let you pass in non-string values for user & password. So I'm calling our `updateProfile` function on newly-created profiles, just after the profile manager is refreshed, to remove any blank fields.

**Also, would like to see the default value for encoding used or leave it out of the yaml**
This one should be fixed. I rearranged the conditions in that loop I added.

**Finally, shouldn't it leave response timeout out of the yaml if value = 0?**
This should also be good now.

**Comment by jellypuno** [comment]
> Should we use this same way of thinking during create?
> I think so (and now Create should work in the new way, not storing empty values). Create is kind of hacky right now because the CliProfileManager.save function does some validation, and won't let you pass in non-string values for user & password. So I'm calling our updateProfile function on newly-created profiles, just after the profile manager is refreshed, to remove any blank fields.

@katelynienaber Let's think of another way and not implement a hack. We can check zowe/cli code as to how they implmented this.

@JillieBeanSim When you say "same way of thinking during create", Are you thinking of re-factoring that process? Code-wise or UX-wise or both? If yes, do you a design that we can look into?

**Comment by katelynienaber** [comment]
@JillieBeanSim Ok, now we have a not-hacky solution :) @jellypuno pointed out that [this option](https://github.com/zowe/imperative/blob/741f24f7246028879317dbd9590b91d7efbdef39/packages/cmd/__tests__/profiles/CliProfileManager.test.ts#L280) could allow passing optional parameters, and it works (at least for me)

Please check again when you have time ^^"

**Comment by JillieBeanSim** [comment]
> 
> @JillieBeanSim When you say "same way of thinking during create", Are you thinking of re-factoring that process? Code-wise or UX-wise or both? If yes, do you a design that we can look into?

@jellypuno I was only thinking of the way it was saved to the yaml, ie. no empty strings and `0` values, not a UI change.

Thanks for looking into my questions and concerns @katelynienaber & @jellypuno I will retest the PR today.

**Comment by JillieBeanSim** [comment]
@katelynienaber the functionality works well now and as expected for create and update of profiles without optional user/pass and for the empty encoding and response timeout values.  I do see this message pop up sometimes after editing a profile that doesn't have the values mentioned.  Not sure but I believe this error is being handled in @jellypuno PR https://github.com/zowe/vscode-extension-for-zowe/pull/1100 I know I ran into it there as well. If this is correct @jellypuno let me know and I will approve this PR

**Comment by zdmullen** [comment]
Merge this after Jelly's PR #1100!

**Comment by katelynienaber** [comment]
@JillieBeanSim I think this issue was fixed in @jellypuno's PR, could you check it out again? I don't see it in mine.

**Review by lauren-li — DISMISSED** [review]
This PR is looking good to me. All unit and expected integration tests are passing. I am able to create and edit profiles without specifying username, password, encoding, and/or timeout, and the yaml is created/updated without empty fields, which is really great. I do sometimes see the error that @JillieBeanSim mentioned in her [comment](https://github.com/zowe/vscode-extension-for-zowe/pull/1120#issuecomment-758738277) above, ... but I can also get it to happen in v1.11.1 the same way (but not reliably). We may want to look into it as a separate issue.

Another issue I encountered while testing, which I also see in 1.11.1 (and thus not a blocker for this PR) is that I get prompted for credentials immediately upon finishing creation or editing of a profile without username/password. I suppose this behavior has been around for a while, but somehow I only just now noticed it, and it feels unexpected since I am not trying to use the profile to access anything yet; I'm just udpating/creating it. Again, this could be investigated as a separate issue. I just wanted to mention it here so others are aware that it isn't a new behavior with this PR.

Thanks @katelynienaber for the fixes in this PR! It was such a pain to deal with empty profile fields when troubleshooting/testing other profile-related PRs, and I am happy to see that getting resolved.

**Comment by jellypuno** [comment]
Maybe its just me but after I created my profile, I edited it and I encountered this error:

image

This is my profile:
```
host: 192.86.32.250
port: 10443
rejectUnauthorized: false
```

**Comment by lauren-li** [comment]
@JillieBeanSim and I have seen an issue in `master` branch, which is not a regression from this PR. However, the fact that this PR works like it's supposed to makes the existing issue much easier for users to encounter. It seems that when clicking a view's "RefreshAll" button, profiles without a user/password field are not refreshed:

ze-refreshAll

In the gif above, the profiles `d`, `ncz`, and `tivNC` all had in common that they were missing user/password fields. Just to be sure, I also tested profiles that had user/password filled in, but lacked encoding/timeout fields (both with good hosts and bad hosts to test active vs inactive), and those profiles were also able to refresh fine.

I also get this message in the debug console when the issue occurs: `TypeError: Cannot read property 'ISession' of undefined at Object.refreshTree`.

We should be sure we can fix this issue in 1.12 before merging this PR in. Again, this PR does not seem to be causing the issue, but it just happens to expose the user to it very easily. (I am still very excited that we are getting rid of empty strings in the profile to mirror what Zowe CLI does, though.)

**Comment by katelynienaber** [comment]
> Another issue I encountered while testing, which I also see in 1.11.1 (and thus not a blocker for this PR) is that I get prompted for credentials immediately upon finishing creation or editing of a profile without username/password.

@lauren-li I will look into this as part of PR #1150

**Comment by katelynienaber** [comment]
@lauren-li I disagree that we should wait to merge this. I'm not convinced we can get another PR finished and reviewed twice by tomorrow. And therefore we would be delaying a bugfix for 3 issues affecting 5 profile fields until the next release, because it reveals 1 issue affecting 2 profile fields.

**Comment by jellypuno** [comment]
Issues that needs to be addressed:

1. Cannot create a profile if I am using SCS
image

2. This appears if I edit a newly created profile: 
image

    the yaml file was updated though.

    image

3. Similar to  @lauren-li comment, the status of the profile is not refreshing.

**Comment by jellypuno** [comment]
For the SCS issue, I think there's something that we need to update in our own SCS. I am not sure what though. I'm hoping @t1m0thyj can help us.... 😊

**Comment by katelynienaber** [comment]
> 1. Cannot create a profile if I am using SCS

Blargh SCS. This still needs fixing...

> 2. This appears if I edit a newly created profile:
>    image

I've fixed this, just copied it out of the old base profiles PR

> 3. Similar to @lauren-li comment, the status of the profile is not refreshing.

@lauren-li @jellypuno I went ahead and fixed this here. Please please can we merge it like this it is such a small change, another PR for this would be overkill 8|

Original code:
`if (sessNode.getSession().ISession !== SessionProfile) {`
New fix:
`if (sessNode.getSession() && sessNode.getSession().ISession !== SessionProfile) {`
Now, it will not try to set the session at all, if the node doesn't already contain one. This is during refresh _only_. It will still update the session if you ex. edit the profile, or enter your creds at a prompt.

> 4. After step# 2 I clicked the magnifying glass and this error appears:
>    image
>    I've enabled validation and I am sure that the hostname is active. I expected that I will be prompted to enter my credentials

I am guessing you have created a profile with user/pass, and then are trying to immediately edit it and remove the user/pass...at least that's what I fixed. I just removed validation in the end of the `editSession` function. When you try to use the profile it will validate you as usual.

**Comment by jellypuno** [comment]
@katelynienaber Once you push your commit, I'll test it again.

**Comment by katelynienaber** [comment]
@jellypuno lol totally forgot to do this

**Comment by lauren-li** [comment]
> @lauren-li I went ahead and fixed this here. Please please can we merge it like this it is such a small change, another PR for this would be overkill 8|

I think it's fine if it is fixed in this PR (which it now appears to be). My intention here wasn't to say that the fix _had_ to be in a separate PR, but that it could be if you didn't want to make any additional changes to this PR, since it was looking pretty good to me at the time I wrote my comment. However, since you've also made updates to address issues @jellypuno found, I will test the PR for those as well now. Thanks @katelynienaber for your hard work on this! I will let you know soon how my testing goes.

**Comment by t1m0thyj** [comment]
> I think issue is [here](https://github.com/zowe/imperative/blob/741f24f7246028879317dbd9590b91d7efbdef39/packages/cmd/src/profiles/CliProfileManager.ts#L408)...imperative's save function is trying to delete the field from an existing profile, if you pass in a profile where the field is missing. Since we are making a new profile, the field doesn't exist and can't be deleted. The causeError I am getting is `Error: Unable to delete credentials. at KeytarCredentialManager.`

@katelynienaber This same issue used to happen in the SCS plugin. It was fixed by updating CredentialManager class to fail silently when it attempts to delete a non-existent secure field. This ensures that the field gets deleted if it exists, and eliminates the unwanted failure.

https://github.com/zowe/zowe-cli-scs-plugin/pull/7/files#diff-7fd952b958aa47583d7349b4404413e8918ccac03052333f1dfcb8990b5af5cfL105-L114

I'd suggest implementing the same fix for ZE's KeytarCredentialManager. If you agree it seems safe, you can remove the error being thrown in `deleteCredentialsHelper` 🙂 
https://github.com/zowe/vscode-extension-for-zowe/blob/f0f93e90e01eb0f725c9c630a2d406d61ae7b51d/packages/zowe-explorer/src/KeytarCredentialManager.ts#L73-L80

**Comment by katelynienaber** [comment]
@t1m0thyj Thank you very much! Merging this fix now. I'm logging the error but it won't fail completely.

**Comment by lauren-li** [comment]
Referencing @jellypuno's comment: https://github.com/zowe/vscode-extension-for-zowe/pull/1120#issuecomment-766713463

> 1. Cannot create a profile if I am using SCS

I don't see this issue when testing now.

> 2. This appears if I edit a newly created profile:

I don't see the error message @jellypuno posted; however, I do see this one when trying to edit a profile that has no username/password (whether newly-created or from a previous session):
Screen Shot 2021-01-25 at 1 17 09 PM

> 3. Similar to @lauren-li comment, the status of the profile is not refreshing.

This appears to have been fixed!

> 4. After step# 2 I clicked the magnifying glass and this error appears:

- I still get these errors in MVS and Jobs when trying to run a search using a profile without credentials. (I also have validation enabled, and was expecting to be prompted for username/password.)
- In USS, when I try to run a search using a profile without credentials, I get a different error:
    Screen Shot 2021-01-25 at 1 11 04 PM

~

Some other odd behavior I saw that I wasn't able to produce on `master` branch:
- After deletion, profiles show up again in the dropdown list of available profiles when clicking the **+** button, even though their corresponding `yaml` file was deleted. (This doesn't happen 100% of the time, but I was able to recreate it fairly frequently, particularly for profiles without credentials.)
    ze-1120-deletep

- Editing a profile (especially one without creds/encoding/timeout) causes it to be removed from the `sessions` list in the `settings.json` file (and therefore, the profile does not load automatically upon session restart). I also was not able to recreate this with 100% consistency, but I was still able to recreate it fairly frequently:
    ze-1120-editp

**Comment by katelynienaber** [comment]
> I don't see the error message @jellypuno posted; however, I do see this one when trying to edit a profile that has no username/password (whether newly-created or from a previous session):

This error is present in master and is related to [this line](https://github.com/zowe/vscode-extension-for-zowe/blob/06a5fcc2bc087c80d00f4ee704616c078274f12b/packages/zowe-explorer/src/abstract/ZoweTreeProvider.ts#L163). It isn't related to the original problem of empty fields being stored improperly and removing this line might cause adverse affects, so I won't fix it here. It doesn't stop you from editing the profile in any case.

> > 1. After step# 2 I clicked the magnifying glass and this error appears:
> 
> * I still get these errors in MVS and Jobs when trying to run a search using a profile without credentials. (I also have validation enabled, and was expecting to be prompted for username/password.)

I am inclined not to fix this one here. It's kind of a big deal XD And it's already going to happen in master if you have any invalid profile and no base profile, it's not caused by the changes in this PR. It occurs because we get to [this point](https://github.com/zowe/vscode-extension-for-zowe/blob/06a5fcc2bc087c80d00f4ee704616c078274f12b/packages/zowe-explorer-api/src/profiles/ZoweExplorerZosmfApi.ts#L82) without ever having prompted for creds. In the original base profiles PR we did [this](https://github.com/zowe/vscode-extension-for-zowe/blob/f29a955740b6c4db74189f501cb994ae3011ebd2/src/api/ZoweExplorerZosmfApi.ts#L147), but to implement something like that would be a huge change that would require lots of testing.

> * In USS, when I try to run a search using a profile without credentials, I get a different error:

I rearranged the search function, so that it tries to retrieve the session from the node _after_ the node's profile is validated. **If you don't have user/pass on the profile, the error above will still occur.** But like I said that error is a biggie and should be dealt with separately. The workaround is to disable validation.

> * After deletion, profiles show up again in the dropdown list of available profiles when clicking the **+** button, even though their corresponding `yaml` file was deleted. (This doesn't happen 100% of the time, but I was able to recreate it fairly frequently, particularly for profiles without credentials.)

This should be fixed.

> * Editing a profile (especially one without creds/encoding/timeout) causes it to be removed from the `sessions` list in the `settings.json` file (and therefore, the profile does not load automatically upon session restart). I also was not able to recreate this with 100% consistency, but I was still able to recreate it fairly frequently:

This should also be fixed.

**Comment by jellypuno** [comment]
@katelynienaber I am still having this error:

image


Steps to reproduce:
1. Create a new profile with invalid URL and without User and Password
2. Edit the said profile and enter the correct URL
3. Click the magnifying glass to search for a dataset

Note: My base profile contains host, port and ru. Same with my service profile. And the URL for both profiles are different

**Comment by jellypuno** [comment]
I can create and edit profiles with token. thank you for that. I tried to submit a job and I encountered this error.

image

**Comment by katelynienaber** [comment]
@jellypuno Can you give it another shot please?

File: packages/zowe-explorer/src/abstract/ZoweTreeProvider.ts
Comment (lauren-li) [inline packages/zowe-explorer/src/abstract/ZoweTreeProvider.ts]: Adding L172 ` this.mHistory.addSession(node.label);` seems to have fixed the issue where the profile was getting removed out of the `settings.json` file after editing, but I'm a little confused what lines 171-172 are meant to be doing now. It looks like L171 just removes the session from `settings.json` and then L172 just adds it back in and cancels L171 out. Could these two lines just be removed?

**Comment by jellypuno** [comment]
Small issue, I think we can just filter this out and stop displaying it.

This appears when I right-click > Update Profile

image

**Comment by jellypuno** [comment]
Error when I tried to add a newly created profile in USS favorites:

image

**Comment by katelynienaber** [comment]
@jellypuno Fixed this...there is still the "profile inactive" error popping up, but the profile is added to favorites. I think this inactive error should be dealt with in another PR, because it occurs all over ZE, not just in profiles operations.

**Comment by lauren-li** [comment]
@katelynienaber What is the expected result if the user has made profile ABC without credentials, does not have a base profile, and uses profile ABC to run a search? I get the set of errors below. I was thinking we should be prompting users for username/password, but I don't get a chance to do that with my new profile:

Screen Shot 2021-01-28 at 12 49 18 PM

**Comment by katelynienaber** [comment]
@lauren-li If you can turn off validation for the profile and get a prompt thereafter...see my post above XD Basically the flow is messed up for validation and it's too big of an unrelated issue to be addressed here.

If you turn off validation and it still won't prompt, let me know, I will try again.

> I am inclined not to fix this one here. It's kind of a big deal XD And it's already going to happen in master if you have any invalid profile and no base profile, it's not caused by the changes in this PR. It occurs because we get to this point without ever having prompted for creds. In the original base profiles PR we did this, but to implement something like that would be a huge change that would require lots of testing.

**Comment by JillieBeanSim** [comment]
@katelynienaber @jellypuno @lauren-li if the user has to do a bunch of work arounds to have this PR in and their profiles to show up and work then I personally don't think this should go out in this release.  I agree the inactive error should be dealt with in another PR but I also believe that this PR and the inactive one will need to go out together possibly as bug fixes in 1.12.1.  This is just my opinion so we don't break users and have to point them to workarounds like disabling validation or requiring them to have a base profile.

**Comment by lauren-li** [comment]
@katelynienaber I'm inclined to agree with @JillieBeanSim on this. This PR is in pretty good shape, I really like it, and like you said, fixing the root cause of the error prompts is a bigger issue. But introducing this feature without addressing those prompts has the potential to degrade the user experience, since they would now be getting errors for something they were able to do without any problems before.

**Comment by JillieBeanSim** [comment]
I may be able to even pitch in with the inactive error/prompting for creds if the field is missing PR as to have these ready for 1.12.1, I was fiddling with it a couple of weeks ago before I jumped in with the filter icon PR

**Comment by katelynienaber** [comment]
@lauren-li @JillieBeanSim It's fine with me if you want to wait on this one, it wasn't promised in the PI goals :D We should bring it up with @zdmullen and @jellypuno as well.

**Review by JillieBeanSim — APPROVED** [review]
This PR works great now with https://github.com/zowe/vscode-extension-for-zowe/pull/1178 added in.  I don't see any of the issues we were seeing before.  Thanks @katelynienaber for doing this work.  All expected tests and checks are passing

**Review by lauren-li — APPROVED** [review]
This looks good to me now! All unit and expected integration tests are passing. Omitting credentials when creating and updating a profile in Zowe Explorer now omits the respective fields instead of creating empty strings. With the merging-in of PR #1178, this works much more seamlessly than before (i.e. no new "error" pop-ups). This PR also appears to be "backwards compatible" with the existing credential-less profiles, in that those profiles are still able to be used, and updating them with empty credentials will just completely remove the blank `user` and `password` fields.

Thanks @katelynienaber for your hard work on this PR!

**Pull request merged**
